File: XamarinStore.iOS/Views/StringSelectionCell.cs
using System;
using MonoTouch.UIKit;
using System.Collections.Generic;
using System.Linq;

namespace XamarinStore
{
	public class StringSelectionCell : UITableViewCell
	{
		public const string Key = "SelectionCell";

		public Action SelectionChanged;

		readonly UIView ViewForPicker;
		readonly StringUIPicker picker;

		public IEnumerable<string> Items {
			get { return picker.Items; }
			set {
				picker.Items = value;

				switch (value.Count ()) {
				case 1:
					DetailTextLabel.TextColor = UIColor.Gray;
					//UserInteractionEnabled = false;
					Accessory = UITableViewCellAccessory.None;
					break;
				default:
					DetailTextLabel.TextColor = UIColor.Black;
					//UserInteractionEnabled = true;
					Accessory = UITableViewCellAccessory.DisclosureIndicator;
					break;
				}
			}
		}

		public string SelectedItem {
			get { return picker.SelectedItem; }
			set { DetailText =picker.SelectedItem = value; }
		}

		public int SelectedIndex {
			get { return picker.SelectedIndex; }
			set { picker.SelectedIndex = value; }
		}

		public string Text {
			get{ return TextLabel.Text; }
			set{ TextLabel.Text = value; }
		}

		public string DetailText {
			get{ return DetailTextLabel.Text; }
			set{ DetailTextLabel.Text = value; }
		}

		public StringSelectionCell (UIView viewForPicker)
			: base (UITableViewCellStyle.Value1, Key)
		{
			SelectionStyle = UITableViewCellSelectionStyle.None;
			TextLabel.TextColor = Color.Purple;
			ViewForPicker = viewForPicker;
			Accessory = UITableViewCellAccessory.DisclosureIndicator;

			picker = new StringUIPicker ();
			picker.SelectedItemChanged += (object sender, EventArgs e) => {
				DetailTextLabel.Text = picker.SelectedItem;
				if (SelectionChanged != null)
					SelectionChanged();
			};

			Items = new string[0];
		}

		public void Tap ()
		{
			// Don't show the picker when we don't have options.
			if (Items.Count () == 1)
				return;

			picker.SelectedIndex = Items.ToList().IndexOf (DetailText);
			picker.ShowPicker (ViewForPicker);
		}
	}
}

File: F#/XamarinStore.iOS/Views/StringSelectionCell.fs
namespace XamarinStore

open System
open MonoTouch.UIKit
open System.Collections.Generic
open System.Linq

[<AllowNullLiteralAttribute>]
type StringSelectionCell(viewForPicker:UIView) as this =
    inherit UITableViewCell(UITableViewCellStyle.Value1, "SelectionCell")

    let Key = "SelectionCell"

    let picker = new StringUIPicker ()

    do picker.SelectedItemChanged <- fun _ -> this.DetailTextLabel.Text <- picker.SelectedItem
                                              this.SelectionChanged()
       picker.Items <- [||]
       this.SelectionStyle <- UITableViewCellSelectionStyle.None
       this.TextLabel.TextColor <- Color.Purple.ToUIColor()
       this.Accessory <- UITableViewCellAccessory.DisclosureIndicator

    member val SelectionChanged = fun ()->() with get,set

    member this.Items
        with get () = picker.Items
        and set value =
            picker.Items <- value

            match value.Count() with
            | 1 -> this.DetailTextLabel.TextColor <- UIColor.Gray
                   this.Accessory <- UITableViewCellAccessory.None
            | _ -> this.DetailTextLabel.TextColor <- UIColor.Black
                   this.Accessory <- UITableViewCellAccessory.DisclosureIndicator

    member this.SelectedItem
        with get () = picker.SelectedItem
        and set value = this.DetailText <- value
                        picker.SelectedItem <- value

    member this.SelectedIndex
        with get () = picker.SelectedIndex
        and set value = picker.SelectedIndex <- value

    member this.Text 
        with get () = this.TextLabel.Text
        and set value = this.TextLabel.Text <- value

    member this.DetailText
        with get () = this.DetailTextLabel.Text
        and set value = this.DetailTextLabel.Text <- value



    member this.Tap () =
        // Don't show the picker when we don't have options.
        if this.Items.Count() <> 1 then
            picker.SelectedIndex <- this.Items.ToList().IndexOf this.DetailText
            picker.ShowPicker viewForPicker
